File: examples/video_det_demo.cc
// End-to-end demo: video file -> VPU hardware decode -> DetectionPipeline.
//
// This is the decode half of the future camera path (a video file standing in
// for live VIN/MIPI capture): each frame is hardware-decoded on the VPU, handed
// to the BPU object detector via DetectionPipeline, and (optionally) written out
// as an annotated JPEG.
//
//   ./video_det_demo <det.hbm> <in.h264|in.h265> [out_dir] [max_frames]
//
// Pipeline per frame:
//   Annex-B access unit --VPU--> NV12 VpImage --cv::cvtColor--> BGR
//     --DetectionPipeline.process--> boxes (original-image px) --draw--> JPEG
//
// The NV12->BGR->(letterbox NV12) hop is a small redundancy we accept to reuse
// the validated BGR-input DetectionPipeline unchanged; a future NV12-direct
// pipeline entry would skip it.

#include <sys/stat.h>

#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>
#include <vector>

#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

#include "bcdl/backend/engine.h"
#include "bcdl/media/video_codec.h"
#include "bcdl/pipeline/detection_pipeline.h"
#include "bcdl/preproc/vp_image.h"

namespace {

std::vector<uint8_t> readFile(const char* path) {
  FILE* fp = std::fopen(path, "rb");
  if (!fp) throw std::runtime_error(std::string("cannot open ") + path);
  std::fseek(fp, 0, SEEK_END);
  long n = std::ftell(fp);
  std::fseek(fp, 0, SEEK_SET);
  std::vector<uint8_t> buf(n > 0 ? static_cast<size_t>(n) : 0);
  if (n > 0 && std::fread(buf.data(), 1, buf.size(), fp) != buf.size()) {
    std::fclose(fp);
    throw std::runtime_error("short read");
  }
  std::fclose(fp);
  return buf;
}

int nalType(uint8_t b, bool h265) { return h265 ? (b >> 1) & 0x3F : b & 0x1F; }
bool isVcl(int t, bool h265) { return h265 ? (t >= 0 && t <= 31) : (t >= 1 && t <= 5); }

// Split an Annex-B byte stream into access units (one coded picture each).
std::vector<std::pair<size_t, size_t>> accessUnits(const std::vector<uint8_t>& s, bool h265) {
  std::vector<std::pair<size_t, size_t>> aus;
  std::vector<size_t> sc;
  for (size_t i = 0; i + 3 <= s.size(); ++i)
    if (s[i] == 0 && s[i + 1] == 0 && s[i + 2] == 1) sc.push_back(i);
  if (sc.empty()) return aus;
  size_t au_begin = sc[0];
  bool au_has_vcl = false;
  for (size_t pos : sc) {
    size_t hdr = pos + 3;
    if (hdr >= s.size()) break;
    bool vcl = isVcl(nalType(s[hdr], h265), h265);
    if (vcl && au_has_vcl) {
      aus.emplace_back(au_begin, pos);
      au_begin = pos;
      au_has_vcl = false;
    }
    if (vcl) au_has_vcl = true;
  }
  aus.emplace_back(au_begin, s.size());
  return aus;
}

// Copy a decoded NV12 VpImage into a tight [h*3/2, w] CV_8UC1 Mat (honouring the
// device strides), then convert to BGR.
cv::Mat nv12ToBgr(const bcdl::VpImage& f) {
  const int w = f.width(), h = f.height();
  cv::Mat nv12(h * 3 / 2, w, CV_8UC1);
  const auto* y = static_cast<const uint8_t*>(f.data());
  const int ys = f.raw().stride;
  for (int r = 0; r < h; ++r) std::memcpy(nv12.ptr(r), y + (size_t)r * ys, w);
  const auto* uv = static_cast<const uint8_t*>(f.raw().uvVirAddr);
  const int us = f.raw().uvStride;
  for (int r = 0; r < h / 2; ++r) std::memcpy(nv12.ptr(h + r), uv + (size_t)r * us, w);
  cv::Mat bgr;
  cv::cvtColor(nv12, bgr, cv::COLOR_YUV2BGR_NV12);
  return bgr;
}

}  // namespace

int main(int argc, char** argv) {
  if (argc < 3) {
    std::fprintf(stderr, "usage: %s <det.hbm> <in.h264|in.h265> [out_dir] [max_frames]\n", argv[0]);
    return 1;
  }
  const char* hbm = argv[1];
  const char* in_path = argv[2];
  const char* out_dir = (argc > 3 && argv[3][0]) ? argv[3] : nullptr;  // "" => no output
  const int max_frames = argc > 4 ? std::atoi(argv[4]) : 0;

  std::string ext;
  if (const char* dot = std::strrchr(in_path, '.')) ext = dot;
  const bool h265 = (ext == ".h265" || ext == ".hevc" || ext == ".265");

  try {
    auto stream = readFile(in_path);
    auto aus = accessUnits(stream, h265);
    if (aus.empty()) {
      std::fprintf(stderr, "no NAL start codes — not an Annex-B stream?\n");
      return 2;
    }
    if (out_dir) ::mkdir(out_dir, 0755);

    bcdl::Engine engine(hbm);
    bcdl::PipelineConfig cfg;
    cfg.detect.num_classes = 80;       // COCO
    cfg.detect.conf_thresh = 0.25f;
    cfg.detect.iou_thresh = 0.45f;
    bcdl::DetectionPipeline pipe(engine, cfg);

    bcdl::VideoDecConfig dcfg;
    dcfg.type = h265 ? HB_VP_VIDEO_TYPE_H265 : HB_VP_VIDEO_TYPE_H264;
    bcdl::VideoDecoder dec(dcfg);

    std::printf("model=%s  inputs=%d outputs=%d  | video=%s (%s, %zu AUs)\n",
                engine.modelName().c_str(), engine.numInputs(), engine.numOutputs(),
                in_path, h265 ? "H.265" : "H.264", aus.size());

    int frames = 0;
    long total_dets = 0;
    double dec_ms = 0, det_ms = 0, cvt_ms = 0, wall_ms = 0;
    bcdl::VpImage nv12;

    // Detect on one decoded frame, and (optionally) write the annotated JPEG.
    auto handleFrame = [&](bcdl::VpImage& f, std::chrono::steady_clock::time_point t0) {
      auto t1 = std::chrono::steady_clock::now();
      cv::Mat bgr = nv12ToBgr(f);
      if (!bgr.isContinuous()) bgr = bgr.clone();
      auto t1b = std::chrono::steady_clock::now();
      cvt_ms += std::chrono::duration<double, std::milli>(t1b - t1).count();

      auto dets = pipe.process(bgr.ptr<uint8_t>(0), bgr.cols, bgr.rows);
      auto t3 = std::chrono::steady_clock::now();
      det_ms += std::chrono::duration<double, std::milli>(t3 - t1b).count();
      wall_ms += std::chrono::duration<double, std::milli>(t3 - t0).count();

      ++frames;
      total_dets += static_cast<long>(dets.size());
      if (frames <= 3 || frames % 30 == 0)
        std::printf("frame %3d: %dx%d  dets=%zu\n", frames, bgr.cols, bgr.rows, dets.size());

      if (out_dir) {
        for (const auto& d : dets) {
          cv::rectangle(bgr, cv::Point((int)d.x1, (int)d.y1),
                        cv::Point((int)d.x2, (int)d.y2), cv::Scalar(0, 255, 0), 2);
          cv::putText(bgr, "c" + std::to_string(d.class_id),
                      cv::Point((int)d.x1, (int)d.y1 - 4), cv::FONT_HERSHEY_SIMPLEX,
                      0.5, cv::Scalar(0, 255, 0), 1);
        }
        char path[1024];
        std::snprintf(path, sizeof(path), "%s/det_%04d.jpg", out_dir, frames);
        cv::imwrite(path, bgr);
      }
    };

    // Feed one AU, then drain EVERY frame the decoder has ready before feeding the
    // next. The VPU decodes on its own threads; feeding faster than we dequeue
    // fills its frame buffers and the next hardware decode stalls (INTERRUPT
    // TIMEOUT). Reorder streams (B-frames, HEVC) also emit frames out of step with
    // the AUs, so the tail only comes out of flush() at end-of-stream.
    for (const auto& [b, e] : aus) {
      if (max_frames && frames >= max_frames) break;
      auto t0 = std::chrono::steady_clock::now();
      while (!dec.feed(stream.data() + b, e - b)) {  // input queue full: drain, retry
        if (!dec.receive(nv12, 5)) break;
        dec_ms += std::chrono::duration<double, std::milli>(
            std::chrono::steady_clock::now() - t0).count();
        handleFrame(nv12, t0);
        t0 = std::chrono::steady_clock::now();
      }
      while (dec.receive(nv12, 0)) {  // drain all frames ready right now
        dec_ms += std::chrono::duration<double, std::milli>(
            std::chrono::steady_clock::now() - t0).count();
        handleFrame(nv12, t0);
        if (max_frames && frames >= max_frames) break;
        t0 = std::chrono::steady_clock::now();
      }
      if (max_frames && frames >= max_frames) break;
    }
    while (frames < max_frames || !max_frames) {  // reorder tail at EOS
      auto t0 = std::chrono::steady_clock::now();
      if (!dec.flush(nv12)) break;
      dec_ms += std::chrono::duration<double, std::milli>(
          std::chrono::steady_clock::now() - t0).count();
      handleFrame(nv12, t0);
    }

    if (frames == 0) {
      std::fprintf(stderr, "FAIL: no frames processed\n");
      return 3;
    }

    // Per-stage time distribution over the whole run. decode/cvt are timed here;
    // preproc/infer/postproc come from the pipeline's own StageProfile so the
    // "detect" cost is broken into its three sub-stages.
    const bcdl::StageProfile& sp = pipe.profile();
    const double wall_f = wall_ms / frames;  // real per-frame wall (decode+cvt+detect)
    auto row = [&](const char* name, double total_ms, const char* unit) {
      const double per_f = total_ms / frames;
      std::printf("  %-14s %7.2f ms/f  %5.1f%%   (%s)\n", name, per_f,
                  wall_f > 0 ? 100.0 * per_f / wall_f : 0.0, unit);
    };
    std::printf("\n=== per-frame time distribution (%d frames, real wall %.2f ms/f) ===\n",
                frames, wall_f);
    row("decode",     dec_ms,          "VPU  hardware decode");
    row("nv12->bgr",  cvt_ms,          "CPU  cvtColor");
    row("  preproc",  sp.preproc_ms,   "CPU  letterbox BGR->NV12");
    row("  infer",    sp.infer_ms,     "BPU  feed + submit/wait");
    row("  postproc", sp.postproc_ms,  "CPU  decode + NMS");
    std::printf("  %-14s %7.2f ms/f  100.0%%\n", "= total wall", wall_f);
    std::printf(
        "FPS: real wall = %.1f  |  detect-only(no decode) = %.1f  |  BPU infer-only = %.1f\n",
        wall_ms > 0 ? frames * 1000.0 / wall_ms : 0.0,
        det_ms > 0 ? frames * 1000.0 / det_ms : 0.0,
        sp.infer_ms > 0 ? frames * 1000.0 / sp.infer_ms : 0.0);
    std::printf("total dets %ld | OK: video -> VPU decode -> DetectionPipeline ran.\n",
                total_dets);
  } catch (const std::exception& e) {
    std::fprintf(stderr, "%s\n", e.what());
    return 2;
  }
  return 0;
}

File: examples/video_det_demo.py
#!/usr/bin/env python3
"""End-to-end video object detection in Python — the counterpart to
examples/video_det_demo_async.cc.

  in.(h264|h265|mp4|mov)  --VPU decode-->  NV12  --cv2-->  BGR
     --AsyncDetectionPipeline (CPU preproc || BPU infer+NMS)-->  boxes
     --draw--> BGR --cv2--> NV12 --VPU encode--> H.264/H.265 elem. stream
     --ffmpeg -c copy (container mux only)--> out.mp4

Every heavy step runs on dedicated silicon: the VPU decodes and encodes, the BPU
detects. ffmpeg is used ONLY as a (de)muxer with `-c copy` — it never decodes or
re-encodes pixels, so the VPU still does all the actual video coding.

  python video_det_demo.py <det.hbm> <in.(h264|h265|mp4|mov)> [out.mp4] [max_frames] [depth]

Notes
- MP4/MOV input is demuxed to an Annex-B elementary stream with ffmpeg first
  (raw .h264/.h265 needs no demux). Output MP4 is produced by muxing the VPU's
  elementary stream with ffmpeg `-c copy`.
- Requires the `bcdl` conda env (bcdl + opencv) and ffmpeg/ffprobe on PATH.
"""

import collections
import os
import subprocess
import sys
import time

import cv2
import numpy as np

import bcdl


# --- Annex-B parsing (mirrors the C++ demo) --------------------------------
def _nal_type(b, h265):
    return ((b >> 1) & 0x3F) if h265 else (b & 0x1F)


def _is_vcl(t, h265):
    return (0 <= t <= 31) if h265 else (1 <= t <= 5)


def access_units(s, h265):
    """Split an Annex-B byte stream into access units (one coded picture each)."""
    sc = [i for i in range(len(s) - 2) if s[i] == 0 and s[i + 1] == 0 and s[i + 2] == 1]
    if not sc:
        return []
    aus, au_begin, au_has_vcl = [], sc[0], False
    for pos in sc:
        hdr = pos + 3
        if hdr >= len(s):
            break
        vcl = _is_vcl(_nal_type(s[hdr], h265), h265)
        if vcl and au_has_vcl:
            aus.append((au_begin, pos))
            au_begin, au_has_vcl = pos, False
        if vcl:
            au_has_vcl = True
    aus.append((au_begin, len(s)))
    return aus


# --- container helpers (ffmpeg used only for muxing, never for coding) -----
def probe_codec(path):
    try:
        out = subprocess.run(
            ["ffprobe", "-v", "error", "-select_streams", "v:0",
             "-show_entries", "stream=codec_name", "-of", "csv=p=0", path],
            capture_output=True, text=True, check=True).stdout.strip()
        return out
    except Exception:
        return ""


def load_annexb(path):
    """Return (elementary_stream_bytes, is_h265). Demux MP4/MOV via ffmpeg."""
    ext = os.path.splitext(path)[1].lower()
    if ext in (".h264", ".264"):
        return open(path, "rb").read(), False
    if ext in (".h265", ".hevc", ".265"):
        return open(path, "rb").read(), True
    # Container (mp4/mov/mkv...): demux to Annex-B without touching the pixels.
    codec = probe_codec(path)
    h265 = codec in ("hevc", "h265")
    bsf = "hevc_mp4toannexb" if h265 else "h264_mp4toannexb"
    print(f"  demux {ext} ({codec or 'unknown'}) -> Annex-B via ffmpeg -c copy")
    data = subprocess.run(
        ["ffmpeg", "-v", "error", "-i", path, "-c:v", "copy",
         "-bsf:v", bsf, "-f", "hevc" if h265 else "h264", "-"],
        capture_output=True, check=True).stdout
    return data, h265


def mux_mp4(elem_path, out_mp4, fps):
    """Wrap the VPU's elementary stream into an MP4 (container only, -c copy)."""
    subprocess.run(
        ["ffmpeg", "-y", "-v", "error", "-r", str(fps), "-i", elem_path,
         "-c", "copy", out_mp4], check=True)


# --- pixel-format conversion ------------------------------------------------
def nv12_to_bgr(flat, w, h):
    """VpImage.to_numpy() NV12 -> (H,W,3) BGR."""
    return cv2.cvtColor(flat.reshape(h * 3 // 2, w), cv2.COLOR_YUV2BGR_NV12)


def bgr_to_nv12_flat(bgr):
    """(H,W,3) BGR -> packed NV12 (W*H*3/2) uint8 [Y then interleaved UV]."""
    h, w = bgr.shape[:2]
    i420 = cv2.cvtColor(bgr, cv2.COLOR_BGR2YUV_I420)  # (H*3/2, W) planar Y,U,V
    y = i420[:h]
    u = i420[h:h + h // 4].reshape(h // 2, w // 2)
    v = i420[h + h // 4:].reshape(h // 2, w // 2)
    uv = np.empty((h // 2, w), dtype=np.uint8)
    uv[:, 0::2] = u   # NV12 = U,V interleaved
    uv[:, 1::2] = v
    return np.ascontiguousarray(np.vstack([y, uv])).reshape(-1)


COLORS = [(0, 255, 0), (0, 200, 255), (255, 128, 0), (200, 0, 255), (0, 128, 255)]


def draw(bgr, dets):
    for d in dets:
        c = COLORS[d.class_id % len(COLORS)]
        cv2.rectangle(bgr, (int(d.x1), int(d.y1)), (int(d.x2), int(d.y2)), c, 2)
        cv2.putText(bgr, f"c{d.class_id} {d.score:.2f}", (int(d.x1), int(d.y1) - 4),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.5, c, 1)


def main(argv):
    if len(argv) < 3:
        print("usage: video_det_demo.py <det.hbm> <in.(h264|h265|mp4|mov)> "
              "[out.mp4] [max_frames] [depth]", file=sys.stderr)
        return 1
    hbm, in_path = argv[1], argv[2]
    out_mp4 = argv[3] if len(argv) > 3 and argv[3] else \
        os.path.splitext(os.path.basename(in_path))[0] + "_det.mp4"
    max_frames = int(argv[4]) if len(argv) > 4 else 0
    depth = int(argv[5]) if len(argv) > 5 else 4
    fps = 30

    stream, h265 = load_annexb(in_path)
    aus = access_units(stream, h265)
    if not aus:
        print("no NAL start codes — not an Annex-B stream?", file=sys.stderr)
        return 2

    engine = bcdl.Engine(hbm)
    cfg = bcdl.PipelineConfig()
    cfg.detect.num_classes = 80          # COCO
    cfg.detect.conf_thresh = 0.25
    cfg.detect.iou_thresh = 0.45
    pipe = bcdl.AsyncDetectionPipeline(engine, cfg, depth)

    dcfg = bcdl.VideoDecConfig()
    dcfg.type = bcdl.VideoType.H265 if h265 else bcdl.VideoType.H264
    dec = bcdl.VideoDecoder(dcfg)

    print(f"model={engine.model_name}  video={in_path} "
          f"({'H.265' if h265 else 'H.264'}, {len(aus)} AUs)  depth={depth} -> {out_mp4}")

    enc = None
    enc_w = enc_h = 0
    elem = bytearray()
    pending = collections.deque()        # decoded BGR frames awaiting their result
    dec_ms = cvt_ms = encprep_ms = enc_ms = 0.0
    total_dets = frames = 0

    elem_path = os.path.splitext(out_mp4)[0] + (".h265" if h265 else ".h264")

    def consume(dets):
        nonlocal enc, enc_w, enc_h, total_dets, encprep_ms, enc_ms
        frame = pending.popleft()
        draw(frame, dets)
        total_dets += len(dets)
        if enc is None:                  # lazily size the encoder (w%32, h%8)
            h, w = frame.shape[:2]
            enc_w, enc_h = (w // 32) * 32, (h // 8) * 8
            ecfg = bcdl.VideoEncConfig()
            ecfg.type = dcfg.type
            ecfg.width, ecfg.height = enc_w, enc_h
            ecfg.framerate = fps
            enc = bcdl.VideoEncoder(ecfg)
        crop = frame[:enc_h, :enc_w]
        t = time.perf_counter()
        vp = bcdl.vp_image_from_nv12(bgr_to_nv12_flat(crop), enc_w, enc_h)  # CPU
        t2 = time.perf_counter()
        # The encoder is a QUEUE, not a function: a packet does not necessarily
        # come out per frame fed in. Feed, then drain to empty; if the input
        # queue is momentarily full, drain and RETRY (at 1080p we generate
        # frames far faster than the VPU encodes them, so this does happen).
        # The tail is flushed after the loop.
        for _ in range(100):
            if enc.feed(vp):
                break
            pkt = enc.receive(5)
            if pkt:
                elem.extend(pkt)
        else:
            raise RuntimeError("VPU encoder never accepted a frame")
        while True:
            pkt = enc.receive(0)
            if pkt is None:
                break
            elem.extend(pkt)                                               # VPU
        enc_ms += (time.perf_counter() - t2) * 1e3
        encprep_ms += (t2 - t) * 1e3

    t_wall = time.perf_counter()
    made = 0

    def submit_frame(vp):
        """Convert one decoded NV12 frame to BGR and push it into the detector."""
        nonlocal made, cvt_ms, frames
        t = time.perf_counter()
        bgr = nv12_to_bgr(vp.to_numpy(), vp.width, vp.height)
        cvt_ms += (time.perf_counter() - t) * 1e3
        pipe.submit(bgr)
        pending.append(bgr)
        made += 1
        if made > depth:
            dets = pipe.next()
            if dets is not None:
                consume(dets)
                frames += 1

    # Feed one AU, then drain every frame the decoder already has. Feeding faster
    # than we dequeue fills the VPU's frame buffers and the next hardware decode
    # stalls (INTERRUPT TIMEOUT); reorder streams also emit frames out of step with
    # the AUs, so the tail only appears via flush() at end-of-stream.
    for b, e in aus:
        if max_frames and made >= max_frames:
            break
        t = time.perf_counter()
        while not dec.feed(bytes(stream[b:e])):     # input queue full: drain, retry
            vp = dec.receive(5)
            dec_ms += (time.perf_counter() - t) * 1e3
            if vp is None:
                break
            submit_frame(vp)
            t = time.perf_counter()
        while (vp := dec.receive(0)) is not None:   # drain all frames ready now
            dec_ms += (time.perf_counter() - t) * 1e3
            submit_frame(vp)
            if max_frames and made >= max_frames:
                break
            t = time.perf_counter()
        if max_frames and made >= max_frames:
            break

    while not max_frames or made < max_frames:      # reorder tail at EOS
        t = time.perf_counter()
        vp = dec.flush()
        dec_ms += (time.perf_counter() - t) * 1e3
        if vp is None:
            break
        submit_frame(vp)

    pipe.finish()
    while (dets := pipe.next()) is not None:
        consume(dets)
        frames += 1
    wall_ms = (time.perf_counter() - t_wall) * 1e3

    # Drain the encoder's tail: packets for frames already fed but not yet
    # emitted. Without this the last frames are silently missing from the mp4.
    if enc is not None:
        while True:
            pkt = enc.flush()
            if pkt is None:
                break
            elem.extend(pkt)

    # Mux the VPU elementary stream into an MP4 (container only).
    with open(elem_path, "wb") as f:
        f.write(elem)
    mux_mp4(elem_path, out_mp4, fps)
    os.remove(elem_path)

    # Per-stage service time (stages overlap; slowest bounds throughput).
    sp = pipe.profile()
    n = max(frames, 1)
    rows = [
        ("decode",    dec_ms / n,               "VPU  hardware decode"),
        ("nv12->bgr", cvt_ms / n,               "CPU  cvtColor (main)"),
        ("preproc",   sp.preproc_per_frame(),   "CPU  letterbox BGR->NV12"),
        ("infer",     sp.infer_per_frame(),     "BPU  feed + submit/wait"),
        ("postproc",  sp.postproc_per_frame(),  "CPU  decode + NMS"),
        ("bgr->nv12", encprep_ms / n,           "CPU  I420->NV12 for encoder"),
        ("encode",    enc_ms / n,               "VPU  hardware encode"),
    ]
    slowest = max(r[1] for r in rows)
    ssum = sum(r[1] for r in rows)
    print(f"\n=== per-stage service time ({frames} frames, stages OVERLAP) ===")
    for name, per_f, unit in rows:
        mark = "  <== bottleneck" if per_f == slowest else ""
        print(f"  {name:<10} {per_f:7.2f} ms/f   ({unit}){mark}")
    print(f"  serial sum = {ssum:.2f} ms/f  ->  overlapped wall = {wall_ms / n:.2f} ms/f  "
          f"(overlap {ssum / (wall_ms / n):.2f}x)")
    print(f"detect+encode {frames} frames | {frames * 1000.0 / wall_ms:.1f} FPS end-to-end "
          f"| total dets {total_dets} -> {out_mp4}")
    print("OK: mp4/h264 -> VPU decode -> BPU detect -> VPU encode -> mp4")
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))
